From 9084536294fd93040e9e20d4573bf94f9c3d72ac Mon Sep 17 00:00:00 2001
Message-Id: <9084536294fd93040e9e20d4573bf94f9c3d72ac.1296065735.git.jdenemar@redhat.com>
From: Eric Blake <eblake@redhat.com>
Date: Tue, 30 Nov 2010 09:00:30 -0700
Subject: [PATCH] qemu: plug memory leak

https://bugzilla.redhat.com/show_bug.cgi?id=672549

src/qemu/qemu_driver.c (qemudShutdown): Free all strings and the
ebtables structure.
* src/libvirt_private.syms (ebtablesContextFree): Export missing
symbol.
* src/util/ebtables.c (ebtablesContextFree): Allow early exit.
(cherry picked from commit 428ea3a6261aa394d2ec465fcc45c45788e5f068)
---
 src/libvirt_private.syms |    1 +
 src/qemu/qemu_driver.c   |   11 ++++++++---
 src/util/ebtables.c      |    2 ++
 3 files changed, 11 insertions(+), 3 deletions(-)

diff --git a/src/libvirt_private.syms b/src/libvirt_private.syms
index 0b5059e..536e14b 100644
--- a/src/libvirt_private.syms
+++ b/src/libvirt_private.syms
@@ -268,6 +268,7 @@ virDomainEventQueueDispatch;
 # ebtables.h
 ebtablesAddForwardAllowIn;
 ebtablesAddForwardPolicyReject;
+ebtablesContextFree;
 ebtablesContextNew;
 ebtablesRemoveForwardAllowIn;
 
diff --git a/src/qemu/qemu_driver.c b/src/qemu/qemu_driver.c
index 411e871..2da74b0 100644
--- a/src/qemu/qemu_driver.c
+++ b/src/qemu/qemu_driver.c
@@ -2243,10 +2243,9 @@ qemudShutdown(void) {
 
     virDomainObjListDeinit(&qemu_driver->domains);
 
-    VIR_FREE(qemu_driver->securityDriverName);
-    VIR_FREE(qemu_driver->logDir);
     VIR_FREE(qemu_driver->configDir);
     VIR_FREE(qemu_driver->autostartDir);
+    VIR_FREE(qemu_driver->logDir);
     VIR_FREE(qemu_driver->stateDir);
     VIR_FREE(qemu_driver->libDir);
     VIR_FREE(qemu_driver->cacheDir);
@@ -2256,9 +2255,15 @@ qemudShutdown(void) {
     VIR_FREE(qemu_driver->vncListen);
     VIR_FREE(qemu_driver->vncPassword);
     VIR_FREE(qemu_driver->vncSASLdir);
-    VIR_FREE(qemu_driver->saveImageFormat);
+    VIR_FREE(qemu_driver->spiceTLSx509certdir);
+    VIR_FREE(qemu_driver->spiceListen);
+    VIR_FREE(qemu_driver->spicePassword);
     VIR_FREE(qemu_driver->hugetlbfs_mount);
     VIR_FREE(qemu_driver->hugepage_path);
+    VIR_FREE(qemu_driver->securityDriverName);
+    VIR_FREE(qemu_driver->saveImageFormat);
+
+    ebtablesContextFree(qemu_driver->ebtables);
 
     if (qemu_driver->cgroupDeviceACL) {
         for (i = 0 ; qemu_driver->cgroupDeviceACL[i] != NULL ; i++)
diff --git a/src/util/ebtables.c b/src/util/ebtables.c
index a6afdf8..27ecebd 100644
--- a/src/util/ebtables.c
+++ b/src/util/ebtables.c
@@ -300,6 +300,8 @@ ebtablesContextNew(const char *driver)
 void
 ebtablesContextFree(ebtablesContext *ctx)
 {
+    if (!ctx)
+        return;
     if (ctx->input_filter)
         ebtRulesFree(ctx->input_filter);
     if (ctx->forward_filter)
-- 
1.7.4.rc3

